pkg/configmap/services/services.go

```diff
@@ -57,6 +68,119 @@ type FileInfo struct {
+	Files []FileInfo `json:"files"`
+}
+
+// NewServicesConfigMapData returns a new servicesCMData. Validates given object contents on creation.
+func NewServicesConfigMapData(services *[]Service, files *[]FileInfo) (*servicesCMData, error) {
+	cmData := &servicesCMData{*services, *files}
+	if err := cmData.validate(); err != nil {
+		return nil, errors.Wrap(err, "unable to create services ConfigMap data object")
+	}
+	return cmData, nil
+}
+
+// Parse converts ConfigMap data into the objects representing a Windows services ConfigMap schema
+// Returns error if the given data is invalid in structure or contents.
+func Parse(data map[string]string) (*servicesCMData, error) {
+	if len(data) != 2 {
+		return nil, errors.New("services ConfigMap should have exactly 2 keys")
+	}
+
+	services := &[]Service{}
+	files := &[]FileInfo{}
+
+	value, ok := data[servicesKey]
+	if !ok {
+		return nil, errors.Errorf("expected `%s` key to exist", servicesKey)
+	}
+	if err := json.Unmarshal([]byte(value), services); err != nil {
+		return nil, errors.Wrapf(err, "`%s` must represent list of Service", servicesKey)
+	}
+
+	value, ok = data[filesKey]
+	if !ok {
+		return nil, errors.Errorf("expected `%s` key to exist", filesKey)
+	}
+	if err := json.Unmarshal([]byte(value), files); err != nil {
+		return nil, errors.Wrapf(err, "`%s` must represent list of FileInfo", filesKey)
+	}
+
+	cmData, err := NewServicesConfigMapData(services, files)
+	if err != nil {
+		return nil, err
+	}
+	return cmData, nil
+}
+
+// validate ensures the given object represents a valid services ConfigMap.
+// The validation ensures bootstrap services always start before controller services according to WICD's expected schema
+func (cmData *servicesCMData) validate() error {
+	if err := validateDependencies(&cmData.Services); err != nil {
+		return err
+	}
+	return validatePriorities(&cmData.Services)
+}
+
+// validateDependencies ensures that no bootstrap service depends on a non-bootstrap service
+func validateDependencies(services *[]Service) error {
+	boostrapServices := []Service{}
+	nonBoostrapServices := []Service{}
+	for _, svc := range *services {
+		if svc.Bootstrap {
+			boostrapServices = append(boostrapServices, svc)
+		} else {
+			nonBoostrapServices = append(nonBoostrapServices, svc)
+		}
+	}
+
+	for _, bootstrapSvc := range boostrapServices {
+		for _, dependency := range bootstrapSvc.Dependencies {
+			for _, nonBootstrapSvc := range nonBoostrapServices {
+				if dependency == nonBootstrapSvc.Name {
+					return errors.Errorf("bootstrap service %s cannot depend on non-boostrap service %s",
+						bootstrapSvc.Name, nonBootstrapSvc.Name)
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// validatePriorities ensures that each service that has the bootstrap flag set as true has a higher priority than all
+// non-bootstrap services. There should be no overlap in the priorities of bootstrap services and controller services.
+func validatePriorities(services *[]Service) error {
```

Comment:
I could be wrong, but aren't slices already a pointer?
passing in a pointer to slice is redundant